keps/20200225-reading-parameter-values-from-a-file.md

```diff
@@ -0,0 +1,157 @@
+types of parameter values.
+
+### Non-Goals
+
+- Make it possible for a user to pass any other command line flags through the file.
+- Provide any templating or other pre-processing capabilities for the parameter value file.
+ - Providing any new way of storage on the back-end.  This is strictly a way of passing parameters to the CLI.
+## Proposal
+
+The idea is to introduce a new command-line flag: `--parameter-file`, in every place that
+currently accepts the `--parameter` flag.
+This flag should require a single argument, interpreted as a path to a YAML file.
+This file should contain parameters a mapping of strings to whatever values are supported as parameter values.
+See the section on [parameter value file schema](#parameter-value-file-schema) for an example.
+
+### User Stories
+
+Currently the only mechanism for passing values to parameters is via the `--parameter`
+command line flag (or its short form `-p`).
+
+This means that the only way to keep track of what parameters were overridden with what
+values is to store the full command line to `kubectl kudo` (or some other form from which
+a command line can be generated with a custom tool).
+
+This presents the following two basic problems.
+
+#### Storing parameter values in version control
+
+Some operators
+([kafka](https://github.com/kudobuilder/operators/blob/master/repository/kafka/operator/params.yaml),
+[cassandra](https://github.com/kudobuilder/operators/blob/master/repository/cassandra/3.11/operator/operator.yaml))
+declare nearly 200 parameters which can be overridden, and we can expect this number to grow.
+
+While it is unlikely that a given operator user will override every single one of these parameters,
+passing them on the command line in large numbers gets inconvenient.
+For example it is hard to provide comments for individual parameter values.
+
+Being able to keep parameter values as a YAML file which can be directly read by
+`kubectl kudo` is a good fit for storing configuration in version control.
+
+#### Passing complex and/or long parameter values
+
+Passing long parameter values through shell command-line, with the need to escape
+shell meta-characters or newlines may become a pain.
+
+These problems will be amplified once KUDO supports parameter value types other than
+plain strings, such as nested dictionaries, etc.
+
+Keeping values in a YAML file means they can be viewed and edited with tools more
+powerful than shell command line editing, such as a text editor with syntax highlighting.
+
+### Implementation Details/Notes/Constraints
+
+#### Parameter value file schema
+
+The top-level element is a mapping from strings which represent parameter names their to values.
+The following example shows a string parameter (supported at the time of writing this KEP)
+as well as an integer and a mapping (to be added in [KEP-24](https://github.com/kudobuilder/kudo/blob/master/keps/0024-parameter-enhancement.md)).
+
+```yaml
```

Comment:
this is NOT valid yaml... this is many yaml docs in 1 yaml file.

Reply:
see https://github.com/kudobuilder/kudo/pull/1020 for more details

Reply:
it is likely best to not use maps as well... this is a kubernetes thing and would be consist with our params file.

What do you think of using the same schema as params.yaml?  this way we have:
1. consistency
2. already have code infra support
3. the extern yaml is an override for the internal yaml which is easy to reason about.

Reply:
> this is NOT valid yaml... this is many yaml docs in 1 yaml file.

Am I missing something? This is a completely valid, simple map, and [YAMLlinter](http://www.yamllint.com/) agrees that this is fine.

About the UX: my understanding is, that all parameters from the above file will be applied? Should we think about applying only a specific parameter from the file? Otherwise, people will end up with many value files for single but complex parameters?

```k kudo update --instance dummy --parameter-file ./values.yaml@MASTER_NODE_AFFINITY```
Maybe something like this?

Reply:
> What do you think of using the same schema as params.yaml?

I'm not sure a proper schema is possible (beyond being a simple map) given individual parameters can be arbitrarily nested (and typed).

Reply:
@kensipe unfortunately I think I have more questions than answers:
- can you explain why you think this is not a single YAML doc? See the [example 2.3 in the "collections" section of the YAML spec](https://yaml.org/spec/1.2/spec.html#id2759963), and [examples 2.7 and 2.9 in the "Structures" section](https://yaml.org/spec/1.2/spec.html#id2760395) for how a single document differs from multiple documents in a file - in particular there is no `---` in this example. In fact, this example is no structurally different from [any old k8s spec file](https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/controllers/nginx-deployment.yaml). I am really puzzled by this comment.
- I read #1020 and its associated issue #1011 which is as terse as it gets :wink: and unfortunately I see no proof for why this would be considered multiple documents.
- I also don't understand the "likely best to not use maps as well... this is a kubernetes thing" argument - as far as I can tell most kubernetes specs are mappings all the way down. I'm all ears though if you could explain why mapping is not a good fit here?
- as for consistency with the structure of `params.yaml` (which is what I assume is you mean by "our params file"), I'm also not convinced. The only advantage of using common syntax that I can see is that they both have something to do with parameters, but just as EBNF grammar for C language does parse itself as a valid C program, I do not see why a parameter value file should resemble a parameter specification file - in fact I would consider it a feature that one is distinguishable from the other at the first sight. Moreover, I do not like the extra verbosity and repetition:
```yaml
foo: bar
baz: quux
```
vs
```yaml
- param: foo
  value: bar
- param: baz
  value: quux
```
- I also don't understand "already have code infra support" - surely parsing a file with a single mapping is just a couple of lines of code? What am I missing?
- Can you please also explain what you mean by "the extern yaml is an override for the internal yaml"? I'm not that familiar with the codebase I'm afraid :-/

Reply:
@zen-dog

> About the UX: my understanding is, that all parameters from the above file will be applied?

Correct.

> Should we think about applying only one parameter from the file? Otherwise, people will end up with many value files for single but complex parameters?
> `k kudo update --instance dummy --parameter-file ./values.yaml@MASTER_NODE_AFFINITY`

Can you explain the use case here? Why would you want to pass a value for just a single parameter from that file? Typically you put in a file all the values that you would like to be specified together. In other words what's the benefit over saying `k kudo update --instance dummy --parameter-file ./values.yaml`? Why would one want to **avoid** passing also all the other parameter values from that file?

To give an example: in my mind this feature is simliar to having a `.bashrc` file. Sure, you have multiple aliases or environment variables defined there, and it could be that sometimes you need `grep` for one of them and set it manually with a customized value. But typically you just source the whole file, because it's faster than typing `@THE_ONE_THING_I_NEED`.

Reply:
> Why would one want to **avoid** passing also all the other parameter values from that file?

I guess, I was thinking about a later point in the operator life-cycle, where a human operator might decide to e.g. tweak `MASTER_NODE_AFFINITY` *only* for already running operator. Now, it's a 15 lines parameter value which makes it challenging to update from the command line. At this point, an operator will have to... copy-paste this particular value into a separate file, change it and then apply it from there? 🤔Now, that I'm writing this down, I'm not sure that my proposal provides a solution in this scenario.

Reply:
Yeah, I don't really like the `./values.yaml@PARAM_NAME`. It introduces another syntax, seems complicated, what if you want multiple parameters from that file, etc.

I think as you can specify multiple `--parameter-file`, this should give the user enough freedom to split their parameter values into different files, if required. 

Additionally, applying only a specific parameter from a file runs a little bit agains the InfrastructureAsCode idea: I want to have everything that is applied to the operator-instance in the `values.yaml`, so I can commit it to my git repository.

Reply:
You're probably right. Still, I was hoping that we would make it easier to apply multiline parameter updates. But we can tackle it later.

Reply:
@zen-dog I see your point. I think for that ad-hoc tweaking use-case something like `kubectl kudo edit-params` (to mimic `kubectl edit`) would make sense.
It would:
1. fetch the parameters from a live instance resource using the mechanism [suggested](https://github.com/kudobuilder/kudo/pull/1364#issuecomment-591311466) earlier by @ANeumann82 
2. write them into a temporary YAML file
3. open the file in your `$EDITOR`
4. you can tweak/add/remove parameters as desired, and save the file+quit editor
5. the CLI applies your changes back to the instance

But let's not add it into this KEP :-D

Reply:
@porridge I agree, this goes beyond this KEP. And I like the `kudo edit` approach - makes it look a lot like what kubectl does. I also thought that one could use other cmd line tools to do the job e.g. to change `weight: 1` to `weight: 2` in the `MASTER_NODE_AFFINITY` parameter *only* one could use a combination of `yq` and `sed`:
```yq r values.yaml MASTER_NODE_AFFINITY | sed '/weight/ s/1/2/'```

Reply:
@porridge the examples provide for yaml are confusing... section 2.3 is on Scalars which doesn't seem associated with this issue.   
I suspect that if you try to create a go example of reading the file format your proposing you will uncover the single doc issue.... 

I don't see value at all in having another format for the same content... I'm a fan of less verbosity where appropriate as well.   In this case... we have defined params for an operator in the params.yaml file (internal to the tgz package).  These represent what is possible and what is defaulted.   It makes sense when overriding these defaults / values to use the same format.   In addition we have existing code to parse this format.  This new format would require new code, new tests for little or no value.  There is lots of win in moving the same format.   What is the justification to adding more code for a similar need... there is also additional cognitive load on the developer to work with parms in multiple way in different situations.

Reply:
@kensipe 
1. Please note I had not said anything about "section 2.3". I said "**Example 2.3** in **section on collections**", i.e. in section 2.1, to which I linked.
1. As for creating code reading the example, [here it is](https://play.golang.org/p/SzqrVVdoaBL) and it works as I expected, so I still do not understand your point. The whole parsing code is 5 lines long, including variable declaration and error handling :-)

I do not agree that this is **the same** content:
- `params.yaml` *declares* what parameters are possible.
- the file under design here *defines* what the parameter *values* are.

Different needs, therefore different format. The former needs multiple properties per parameter, therefore each needs a name, and there is not much difference between keeping the name as a `name` field in a list, vs as a key in a map.

OTOH In the latter case there is **just** the key and the (user-defined) value - a YAML mapping is a perfect fit. I really do not see the need to shoehorn this into using a format similar to the one for `params.yaml`. Note that it will *not* be "the same format" - in fact the only common thing between `params.yaml` and this one is the `name` field. Every other field that `params.yaml` has is unnecessary, and at the same time only other field needed here - `value` is missing!

Reply:
it is a subset of the same content